### ISSUE TITLE
Customed Metric Function executes Twice?

### ISSUE THREAD

**Issue by asherzhao8** [trevorstephens/gplearn#208]
Hello,

I custom a metric function, and print the metric result every time when a new formula is generated. But I find the metric function will be executed twice each time, just like the image below. It doubles the training time, so can you help me figure out the reason?

Thank you very much!
![1](https://user-images.githubusercontent.com/61578765/96964608-5a790100-153d-11eb-8f71-6892d36f93f6.jpg)

**Comment by trevorstephens** [comment]
Sorry but I don't understand what your problem is, most of the output is missing. Have you removed/edited some of the output?

As I cannot run any code, recreate your problem, understand what commands you have run, there is very little I can do to help without more information.

**Comment by fenghuolanshan123** [comment]
> Sorry but I don't understand what your problem is, most of the output is missing. Have you removed/edited some of the output?
> 
> As I cannot run any code, recreate your problem, understand what commands you have run, there is very little I can do to help without more information.

 @trevorstephens ，Sorry for my  expression.
I just add a ‘print’ to the customed metric function. I want to see the metric result each time. Then I find that the result printed every two times is the same, just like the figure above attached in my question. The first one is 0.01647, and the second is still 0.01647. As for why there is no output, it is because the first generation of the model is still in the process of training,  and I didn't change any other parts of the model.

![metric](https://user-images.githubusercontent.com/61578765/97149334-957e6d00-17a7-11eb-9244-db1e725b7898.jpg)

**Comment by trevorstephens** [comment]
Without a reproducible example it's near impossible for me to look into this.

It's possible this is the out of bag fitness being calculated? I won't comment further without a full reproducible example of the problem you wish to solve. Impossible to know what you are doing

**Comment by trevorstephens** [comment]
Try crafting a toy dataset with your problem using the sklearn blobs or something like that, otherwise I cannot look into what is happening.

**Comment by fenghuolanshan123** [comment]
`import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
from sklearn.utils import check_random_state
from gplearn.fitness import make_fitness
from gplearn.genetic import SymbolicTransformer

rng = check_random_state(0)
boston = load_boston()
perm = rng.permutation(boston.target.size)
boston.data = boston.data[perm]
boston.target = boston.target[perm]

def _my_metric(y, y_pred, w):
    value = np.sum(np.abs(y) + np.abs(y_pred))
    print(value)
    return value
my_metric = make_fitness(function=_my_metric, greater_is_better=False)

function_set = ['add', 'sub', 'mul', 'div',
                'sqrt', 'log', 'abs', 'neg', 'inv',
                'max', 'min']
gp = SymbolicTransformer(generations=5, population_size=200,
                         hall_of_fame=100, n_components=10,
                         function_set=function_set, metric=my_metric,
                         parsimony_coefficient=0.0005,
                         max_samples=0.9, verbose=1,
                         random_state=0, n_jobs=3)
gp.fit(boston.data[:300, :], boston.target[:300])`

@trevorstephens .It seems that I have encountered a new problem. Why did I add ‘print(value)’ in the custom metric function, but did not print the metric result in the process of training the model?

**Comment by hwulfmeyer** [comment]
```
gp = SymbolicTransformer(generations=3, population_size=10,
                        hall_of_fame=10, n_components=10,
                        function_set=function_set, metric=my_metric,
                        parsimony_coefficient=0.0005,
                        max_samples=0.9, verbose=1,
                        random_state=0, n_jobs=1)

```
My output looks like this:
```
    |   Population Average    |             Best Individual              |
---- ------------------------- ------------------------------------------ ----------
 Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
69477.63277052
69477.63277052
248517335.35458386
248517335.35458386
6889.867984966956
6889.867984966956
11234.944530059192
11234.944530059192
6642.3
6642.3
6643.937139416466
6643.937139416466
12082.188280356442
12082.188280356442
10773.009932388375
10773.009932388375
7678.773066514686
7678.773066514686
9428.260977538152
9428.260977538152
   0    12.70      2.48658e+07        4           6642.3           6642.3      0.05s
6642.3
6642.3
6942.3
6942.3
6664.3
6664.3
6664.3
6664.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
   1     4.60           6676.7        6           6642.3           6642.3      0.02s
6664.3
6664.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
6664.3
6664.3
6642.3
6642.3
6664.3
6664.3
6664.3
6664.3
6642.3
6642.3
   2     2.80           6651.1        3           6642.3           6642.3      0.00s
```

The reason the output is doubled is because you activated sampling with `max_samples=0.9`. This causes that there is an OOB fitness.
```
        program.raw_fitness_ = program.raw_fitness(X, y, curr_sample_weight)
        if max_samples < n_samples:
            # Calculate OOB fitness
            program.oob_fitness_ = program.raw_fitness(X, y, oob_sample_weight)
```

If you set `max_samples=1.0` there is no double output.
```

    |   Population Average    |             Best Individual              |
---- ------------------------- ------------------------------------------ ----------
 Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
69477.63277052
248517335.35458386
6889.867984966956
11234.944530059192
6642.3
6643.937139416466
12082.188280356442
10773.009932388375
7678.773066514686
9428.260977538152
   0    12.70      2.48658e+07        4           6642.3              N/A      0.02s
6642.3
6942.3
6664.3
6664.3
6642.3
6642.3
6642.3
6642.3
6642.3
6642.3
   1     4.60           6676.7        6           6642.3              N/A      0.02s
6664.3
6642.3
6642.3
6642.3
6642.3
6664.3
6642.3
6664.3
6664.3
6642.3
   2     2.80           6651.1        3           6642.3              N/A      0.00s
```

**Comment by trevorstephens** [comment]
Thx @hwulfmeyer thought it was probably the OOB fitness, appreciate you helping out! 😄

**Comment by fenghuolanshan123** [comment]
@hwulfmeyer I got it, thank you very much! By the way, the function was created casually, i'll never use it. haha!
